File: crystal-ball/cb.py
from math import sqrt, floor

def linear_search(floors, i):
  while i < len(floors):
    if floors[i] == True:
      return i
    i += 1

def crystal_ball(floors) -> int:
  low, sr = 0, floor(sqrt(len(floors)))
  while sr <= len(floors):
    if floors[sr] == True:
      return linear_search(floors, low)
    low = sr
    sr += sr
  return -1

print(crystal_ball([False,False,False,False,False,False,False,False,False,True,True,True,True,True,True,True,True,True]))
print(crystal_ball([False,False,False,False,False,False,False,False,False]))
print(crystal_ball([False,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True,True]))

File: crystal-ball/cb.ts
function crystalBall(floors: boolean[]): number{
  const sr = Math.floor(Math.sqrt(floors.length))
  let [low, high] = [0, sr]
  do {
    if (floors[high] === true){
      return linearSearch(floors, low)
    }
    low = sr
    high += sr
  } while (sr < high)
  return -1
}

function linearSearch(floors: boolean[], low: number): number {
  do {
    if (floors[low] === true) return low
    else low++
  } while (true)
}

console.log(crystalBall([false,false,false,false,false,false,false,false,false,true,true,true,true,true,true,true,true,true]))
console.log(crystalBall([false,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true,true]))